lib/mempool/mempool.c

```diff
@@ -0,0 +1,402 @@
+			free_l = i;
+		}
+	}
+
+	if (alloc_l < 0 || free_l < 0) {
+		*data_p = NULL;
+		return -ENOMEM;
+	}
+
+	/* Iteratively break the smallest enclosing block... */
+	data = block_alloc(p, free_l, lsizes[free_l]);
+
+	if (!data) {
+		/* This can happen if we race with another allocator.
+		 * It's OK, just back out and the timeout code will
+		 * retry.  Note mild overloading: -EAGAIN isn't for
+		 * propagation to the caller, it's to tell the loop in
+		 * k_mem_pool_alloc() to try again synchronously.  But
+		 * it means exactly what it says.
+		 *
+		 * This doesn't happen for user mode memory pools as this
+		 * entire function runs with a semaphore held.
+		 */
+		return -EAGAIN;
+	}
+
+	for (from_l = free_l; from_l < alloc_l; from_l++) {
+		data = block_break(p, data, from_l, lsizes);
+	}
+
+	*level_p = alloc_l;
+	*block_p = block_num(p, data, lsizes[alloc_l]);
+	*data_p = data;
+
+	return 0;
+}
+
+void _sys_mem_pool_block_free(struct sys_mem_pool_base *p, u32_t level,
+			      u32_t block)
+{
+	size_t lsizes[p->n_levels];
+	int i;
+
+	/* As in _sys_mem_pool_block_alloc(), we build a table of level sizes
+	 * to avoid having to store it in precious RAM bytes.
+	 * Overhead here is somewhat higher because block_free()
+	 * doesn't inherently need to traverse all the larger
+	 * sublevels.
+	 */
+	lsizes[0] = _ALIGN4(p->max_sz);
+	for (i = 1; i <= level; i++) {
+		lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+	}
+
+	block_free(p, level, lsizes, block);
+}
+
+
+/*
+ * Functions specific to user-mode blocks
+ */
+
+struct sys_mem_pool *sys_mem_pool_create(char *buf, size_t size,
+					 size_t minsz, size_t maxsz,
+					 struct k_mutex *mutex)
+{
+	size_t buf_size, lvls_size, bits_size;
+	struct sys_mem_pool *p;
+	struct sys_mem_pool_lvl *lvls;
+	u32_t n_max;
+
+	lvls_size = sizeof(struct sys_mem_pool_lvl) * _MPOOL_LVLS(maxsz, minsz);
+	buf_size = size - (sizeof(struct sys_mem_pool) + lvls_size);
+
+	p = (struct sys_mem_pool *)buf;
+	lvls = (struct sys_mem_pool_lvl *)(buf + sizeof(struct sys_mem_pool));
+	buf = buf + sizeof(struct sys_mem_pool) + lvls_size;
+
+	/* Iteratively compute space for both the actual memory blocks and
+	 * associated housekeeping structures, based on available remaining
+	 * size.
+	 *
+	 * XXX: A polynomial equation for _MPOOL_BITS_SIZE should make this
+	 * unnecessary.
+	 */
+	for (n_max = buf_size / maxsz; ; n_max--) {
+		__ASSERT_NO_MSG(n_max > 0);
+		bits_size = _MPOOL_BITS_SIZE(maxsz, minsz, n_max);
+		if (bits_size + _ALIGN4(maxsz * n_max) <= buf_size) {
+			/* Everything fits, we're done */
+			break;
+		}
+	}
+
+	p->base.buf = buf;
+	p->base.max_sz = maxsz;
+	p->base.n_max = n_max;
+	p->base.n_levels = _MPOOL_LVLS(maxsz, minsz);
+	p->base.max_inline_level = 0; /* necessary? */
+	p->base.levels = lvls;
+	p->base.flags = SYS_MEM_POOL_USER;
+	p->mutex = mutex;
+
+	_sys_mem_pool_base_init(&p->base);
+
+	return p;
+}
+
+struct sys_mem_pool_block {
+	struct sys_mem_pool *pool;
+	u32_t level : 4;
+	u32_t block : 28;
+};
```

Comment:
FWIW: there's a similar trick being played (putting the metadata into the allocated block) in k_malloc/k_free that you might consider unifying in some way.  Johan cleaned that up in 1a8a8d901 such that it just stores a k_mem_block_id by value instead of unpacking the fields.  Would save a few code bytes to share the trick, I suspect.

To clarify: it can't be identical because you have to store the pool pointer here and k_malloc can get it implicitly.  But something could be cleaned up I think.